Source/Scene/ShadowMapShader.js

```diff
@@ -15,79 +15,175 @@ define([
+        var sources = fs.sources.slice(0);
+
+        var positionVaryingName = findPositionVarying(fs);
+        var hasPositionVarying = defined(positionVaryingName);
+        if (!hasPositionVarying) {
+            positionVaryingName = 'v_positionEC';
+        }
+
+        if (isTerrain && !terrainHasPositionVarying(defines)) {
+            defines.push('ENABLE_DAYNIGHT_SHADING');
+        }
+
+        var length = sources.length;
+        for (var i = 0; i < length; ++i) {
+            sources[i] = ShaderSource.replaceMain(sources[i], 'czm_shadow_cast_main');
+        }
+
+        var fsSource = '';
+
+        if (isPointLight) {
+            if (!hasPositionVarying) {
+                fsSource += 'varying vec3 v_positionEC; \n';
+            }
+            fsSource += 'uniform vec4 u_shadowMapLightPositionEC; \n';
+        }
 
         if (opaque) {
-            fs =
-                'varying vec3 ' + positionVaryingName + ';\n' +
+            fsSource +=
                 'void main() \n' +
                 '{ \n';
         } else {
-            fs = ShaderSource.replaceMain(fs, 'czm_shadow_main');
-            fs +=
+            fsSource +=
                 'void main() \n' +
                 '{ \n' +
-                '    czm_shadow_main(); \n' +
+                '    czm_shadow_cast_main(); \n' +
                 '    if (gl_FragColor.a == 0.0) { \n' +
                 '       discard; \n' +
                 '    } \n';
         }
 
         if (isPointLight) {
-            fs +=
+            fsSource +=
                 'float distance = length(' + positionVaryingName + '); \n' +
                 'distance /= u_shadowMapLightPositionEC.w; // radius \n' +
                 'gl_FragColor = czm_packDepth(distance); \n';
-        } else if (usesDepthTexture) {
-            fs += 'gl_FragColor = vec4(1.0); \n';
+        } else if (useDepthTexture) {
+            fsSource += 'gl_FragColor = vec4(1.0); \n';
         } else {
-            fs += 'gl_FragColor = czm_packDepth(gl_FragCoord.z); \n';
+            fsSource += 'gl_FragColor = czm_packDepth(gl_FragCoord.z); \n';
         }
 
-        fs += '} \n';
+        fsSource += '} \n';
 
-        if (isPointLight) {
-            if (!hasPositionVarying) {
-                fs = 'varying vec3 v_positionEC; \n' + fs;
-            }
-            fs = 'uniform vec4 u_shadowMapLightPositionEC; \n' + fs;
-        }
+        sources.push(fsSource);
 
-        return fs;
+        return new ShaderSource({
+            defines : defines,
+            sources : sources
+        });
     };
 
-    ShadowMapShader.createShadowReceiveVertexShader = function(vs, frameState) {
-        return vs;
+    ShadowMapShader.createShadowReceiveVertexShader = function(vs, isTerrain) {
+        var defines = vs.defines.slice(0);
+        var sources = vs.sources.slice(0);
+
+        if (isTerrain && !terrainHasPositionVarying(defines)) {
+            defines.push('ENABLE_DAYNIGHT_SHADING');
+        }
```

Comment:
Side note, it was my bad for re-using both the `ENABLE_DAYNIGHT_SHADING` and `ENABLED_VERTEX_LIGHTING` defines in order to generate positions and normals for shadows. I think it's clearer if `GlobeVS` looks something like:

```glsl
#if defined(ENABLE_VERTEX_LIGHTING) || defined(GENERATE_POSITIONS_AND_NORMALS)
    v_positionEC = (u_modifiedModelView * vec4(position, 1.0)).xyz;
    v_positionMC = position3DWC;                                 // position in model coordinates
    v_normalMC = czm_octDecode(encodedNormal);
    v_normalEC = czm_normal3D * v_normalMC;
#elif defined(SHOW_REFLECTIVE_OCEAN) || defined(ENABLE_DAYNIGHT_SHADING) || defined(GENERATE_POSITIONS)
    v_positionEC = (u_modifiedModelView * vec4(position, 1.0)).xyz;
    v_positionMC = position3DWC;                                 // position in model coordinates
#endif
```